## Preface



We need some particular versions of the following modules;



In [1]:
!pip install -r requirements.txt

You are using pip version 9.0.3, however version 20.0.2 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.


Add a string for your username here:



In [2]:
import numpy as np
user = "mattarena"

In [3]:
from __future__ import division
def mean(a):
    return sum(a) / len(a)

#print map(mean, zip(*a))

## USDA Food Central DataBase



The USDA maintains a database of nutritional information, where
  different kinds of food are identified by an FDC number.  They do
  not provide any data on prices.  

To look up nutritional information, use api provided by the USDA at
[https://fdc.nal.usda.gov/](https://fdc.nal.usda.gov/).   You&rsquo;ll need to sign up for a
free api key (see directions on page), then add that key here:



In [4]:
import fooddatacentral as fdc
apikey = {'mattarena':"inIyO1begWSRqsYtxS7m6p09PSyq7Qiw7fxzV2qN"}
fdc.search(apikey[user], "vegetable oil")

allHighlightFields  \
0            <b>Ingredients</b>: SOYBEAN <em>OIL</em>   
1           <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
2           <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
3           <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
4           <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
5           <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
6           <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
7           <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
8           <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
9           <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
10          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
11          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
12  <b>Ingredients</b>: ALL <em>VEGETABLE</em> SOY...   
13          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
14          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
15          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
16           <b>Ingredients</b>: SOYBEAN <em>OIL</em>   
17          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
18          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
19          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
20          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
21          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
22     <b>Ingredients</b>: 100% SOYBEAN <em>OIL</em>.   
23          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
24          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
25          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
26          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
27  <b>Brand Owner</b>: LIFE <em>OIL</em><br/><b>I...   
28          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
29        <b>Ingredients</b>: SUNFLOWER <em>OIL</em>.   
30          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
31          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
32         <b>Ingredients</b>: SOYBEAN <em>OIL</em>.    
33          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
34          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
35          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
36          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
37          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
38          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
39          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
40          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
41          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
42          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
43          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
44          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
45          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
46          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
47          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
48          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   
49          <b>Ingredients</b>: SOYBEAN <em>OIL</em>.   

                     brandOwner dataType    description   fdcId       gtinUpc  \
0   L & J General International  Branded  VEGETABLE OIL  600540  099336006404   
1                  Hy-Vee, Inc.  Branded  VEGETABLE OIL  442919  075450048025   
2             Stratas Foods LLC  Branded  VEGETABLE OIL  475061  751884910434   
3                SMART & SIMPLE  Branded  VEGETABLE OIL  372064  751884912162   
4            Harris-Teeter Inc.  Branded  VEGETABLE OIL  571915   07203632012   
5            Harris-Teeter Inc.  Branded  VEGETABLE OIL  465773  072036320285   
6                 CLOVER VALLEY  Branded  VEGETABLE OIL  690552  072071153770   
7           Lou Ana Foods, Inc.  Branded  VEGETABLE OIL  442753  073391000249   
8                       SPARTAN  Branded  VEGETABLE OIL  383645  011213172262   
9                      SHURFINE  Branded  VEGETABLE OIL  695326  015400840606   
10          Hannaford Bros. Co.  Branded  VEGETABLE OIL  636852  041268195589   
11        Unified Grocers, Inc.  Branded  VEGETABLE OIL  741242  041380111061   
12                CLOVER VALLEY  Branded  VEGETABLE OIL  616402  059872000

### Looking up foods



I&rsquo;ve written a little module `fooddatacentral` with the methods

-   `search`
-   `nutrients`
-   `units`
-   `ingredients`

We&rsquo;ll go through these in sequence.



### FDC Search



Here&rsquo;s a little code to help look up FDC codes for foods of
different descriptions.



In [4]:

#df_tofu=fdc.search(apikey[user],"tofu")
#df_tofu
#for i in df_tofu.
#select row, then column
#int(df_tofu.loc[df_tofu["brandOwner"] == ""].iloc[0, 5])


vegetarian=["tofu", "lentils", "beans", "eggs", "carrots", "oranges", "apples", "oranges", "beets", "celery", "onions", "brocolli", "lemon", "veggie broth", "rice", "wheat flour", "whole wheat flour", "yogurt", "cheddar cheese", "swiss cheese", "jack cheese", "vegetable oil", "olive oil",
"margarine", "sugar", "potatoes", "pasta", "Milk",
"Loaf of Bread", "salt", "brown rice", "wheat bread"]

codes1=[]
for i in vegetarian: 
    df=fdc.search(apikey[user], i)
    food_code=df.loc[df["description"]==i.upper()]
    food_code_list=food_code["fdcId"].tolist()
    codes.append(food_code_list)
print(codes)




    



[[496446, 411177, 514921, 388749, 498775, 391880, 392410], [539527, 522842, 517834, 519343, 656494, 597467, 528168, 357866, 559825, 387766, 487576, 547285, 407778, 728714, 698762, 410162, 671414, 671432, 744306, 645486, 644726, 549819, 543075, 441150, 710520, 475147, 529168, 474561, 555838, 637194, 472066, 364037, 674396, 707036, 457663, 577538, 711366, 523151, 489279, 553468], [548596, 444892], [577532, 666802, 578455, 520005, 578012, 578148, 445613, 578111, 579128, 511826, 578453, 442160, 511760], [450701, 626538, 746412, 658616, 657468, 518787, 411443, 465064, 475485, 487520, 487542, 487626, 524749, 746422, 515272], [578523, 452141], [577833, 524455, 592260, 578270, 577083, 577805, 555422, 539160, 577507, 447524, 448085, 577589, 566572, 578033, 487785, 539161, 472842, 577803], [578523, 452141], [384502, 361663, 465327, 452096], [447185, 528058, 451477, 539577], [], [], [410353], [719162], [356554, 485474, 381062, 356582, 450892, 360217, 490609, 522870, 504855], [390092, 602621, 5676

In [5]:
codes2=[]

vegan=["tofu", "lentils", "beans",
"salt", "carrots", "apples", "oranges",
"beets", "celery", "onions", "brocolli", 
"lemon", "veggie broth", "rice", "wheat flour",
"whole wheat flour", "vegetable oil", "olive oil", "margarine",
"sugar", "potatoes", "pasta", "soy Milk",
"White Bread", "Wheat Bread", "Almond milk"]

#oranges
vegan_codes=[]
for i in vegan: 
    df=fdc.search(apikey[user], i)
    food_code=df.loc[df["description"]==i.upper()]
    food_code_list=food_code["fdcId"].tolist()
    vegan_codes.append(food_code_list)
print(vegan_codes)


[[514921, 411177, 496446, 388749, 498775, 391880, 392410], [559825, 656494, 597467, 528168, 357866, 517834, 539527, 519343, 522842, 728714, 698762, 387766, 487576, 407778, 410162, 547285, 644726, 671414, 671432, 744306, 645486, 441150, 710520, 543075, 549819, 475147, 529168, 474561, 637194, 472066, 555838, 364037, 674396, 707036, 457663, 577538, 711366, 553468, 523151, 489279], [548596, 444892], [556318, 596928, 509410, 448728, 401059, 704294, 394689, 561079, 574441, 568630, 361591, 450069, 623878, 382131, 472490, 450276, 450158, 390965, 777928, 580779, 451908], [450701, 626538, 746412, 658616, 657468, 518787, 411443, 465064, 475485, 487520, 487542, 487626, 524749, 746422, 515272], [577833, 524455, 592260, 577083, 577805, 578270, 555422, 539160, 577507, 447524, 448085, 577589, 566572, 578033, 487785, 539161, 472842, 577803], [578523, 452141], [384502, 361663, 465327, 452096], [447185, 528058, 451477, 539577], [], [], [410353], [719162], [356554, 485474, 356582, 450892, 381062, 360217, 

In [ ]:
Omnivore=["tofu", "lentils", "beans",
"salt", "carrots", "apples", "oranges",
"beets", "celery", "onions", "brocolli", 
"lemon", "veggie broth", "rice", "wheat flour",
"whole wheat flour", "vegetable oil", "olive oil", "margarine",
"sugar", "potatoes", "pasta", "soy Milk",
"White Bread", "Wheat Bread", "Almond milk"]

#oranges
codes=[]
for i in vegan: 
    df=fdc.search(apikey[user], i)
    food_code=df.loc[df["description"]==i.upper()]
    food_code_list=food_code["fdcId"].tolist()
    codes.append(food_code_list)
print(codes)

In [14]:
code=[[2 , 3] , [4]]
code2=[1, 2, 3]
len(code)
type(map(sum, zip(code)))

map

### FDC Nutrients



In [8]:
import fooddatacentral as fdc

df_tofu=fdc.nutrients(apikey[user],594272)
df_tofu["Quantity"].tolist()

[0.0,
 48.28,
 0.0,
 241.0,
 0.0,
 1.72,
 0.0,
 0.0,
 3.4,
 110.0,
 2.48,
 4.138,
 6.9,
 0.352,
 448.0,
 6.9,
 1.0,
 3.45,
 0.0,
 0.0]

Once we know the `fdc_id` of a particular food we can look up a
variety of information on it.  We start with nutrients



In [30]:
import pandas as pd 

vegan_dictionary={}
nutrients_dict={}
full_units_view=[]
#iterates list to take one food at a time
for n,i in enumerate(vegan_codes):
    #iterates each code so we select one code at a time
    for j in i:
        #pulls list of nutrients quantities
        nutrients=fdc.nutrients(apikey[user],j)["Quantity"]
        #now nutrients is list of lists
        nutrients_dict[j] = nutrients
    #for m in np.arange(0, len(nutrients)):
     #   averages=sum(nutrients[m][j])/len(nutrients_list)
    #if len(nutrients_list)==0:
       # vegan_codes=vegan_codes.drop(nutrients_list)
        #print(vegan_codes[i] + "will have to be looked up individually")
    #full_units_view.append(full_mean)

#full_units_view
    df=pd.DataFrame(nutrients_dict)     
    DF[vegan[n]]=df.T.mean()
    
vegan_dataframe = pd.DataFrame(DF)

In [31]:
vegan_dataframe

tofu     lentils       beans  \
Calcium, Ca                               121.142857   38.021277   40.265306   
Carbohydrate, by difference                 2.198571   47.126809   46.961429   
Cholesterol                                 0.000000    0.297872    0.285714   
Copper, Cu                                       NaN         NaN         NaN   
Energy                                     83.428571  233.369565  232.645833   
Fatty acids, total monounsaturated          1.170000    0.731250    0.731250   
Fatty acids, total polyunsaturated          2.572000    1.607500    1.607500   
Fatty acids, total saturated                0.494286    0.073617    0.070612   
Fatty acids, total trans                    0.000000    0.000000    0.000000   
Fiber, insoluble                                 NaN         NaN         NaN   
Fiber, soluble                                   NaN         NaN         NaN   
Fiber, total dietary                        0.871429   19.938298   19.732653   
Folate, DFE                                      NaN         NaN         NaN   
Folate, total                                    NaN         NaN         NaN   
Folic acid                                       NaN         NaN         NaN   
Iodine, I                                        NaN         NaN         NaN   
Iron, Fe                                    1.698571    6.341915    6.261429   
Magnesium, Mg                                    NaN         NaN         NaN   
Niacin                                           NaN         NaN         NaN   
Phosphorus, P                                    NaN         NaN         NaN   
Potassium, K                              196.500000  597.300000  597.300000   
Protein                                     8.631429   20.662128   20.377551   
Riboflavin                                       NaN         NaN         NaN   
Sodium, Na                                 14.000000   27.446809   33.918367   
Sugars, added                                    NaN         NaN    0.800000   
Sugars, total including NLEA                0.000000    1.346444    1.321915   
Thiamin                                          NaN         NaN         NaN   
Total lipid (fat)                           4.230000    1.088511    1.113061   
Total sugar alcohols                             NaN         NaN         NaN   
Vitamin A, IU                               0.000000    0.000000    0.000000   
Vitamin B-12                                     NaN         NaN         NaN   
Vitamin B-6                                      NaN         NaN         NaN   
Vitamin C, total ascorbic acid              0.000000    2.205882    2.142857   
Vitamin D (D2 + D3), International Units         NaN    0.000000    0.000000   
Vitamin E                                        NaN         NaN         NaN   
Zinc, Zn                                         NaN         NaN         NaN   

                                                  salt      carrots  \
Calcium, Ca                                  39.460000    36.800000   
Carbohydrate, by difference                  32.873000    28.676235   
Cholesterol                                   0.269231     0.212121   
Copper, Cu                                         NaN          NaN   
Energy                                      161.840580   139.500000   
Fatty acids, total monounsaturated            0.585000     0.585000   
Fatty acids, total polyunsaturated            1.286000     1.286000   
Fatty acids, total saturated                  0.066538     0.052424   
Fatty acids, total trans                      0.000000     0.000000   
Fiber, insoluble                                   NaN          NaN   
Fiber, soluble                                     NaN          NaN   
Fiber, total dietary                         18.958824    15.256061   
Folate, DFE                                        NaN          NaN   
Folate, total                                      NaN          NaN   
Folic acid            

## Data on Prices



Here&rsquo;s an effort to describe some different kinds of food, along with
data on food prices.  This is all just based on a trip I took to the
grocery store, except that I&rsquo;ve used the USDA database to look up FDC
numbers.  

Also note that some kinds of foods need to have unit weights (in
grams) supplied under &ldquo;Units&rdquo;; e.g., extra large eggs are taken to
each weigh 56g.  These conversions can also often be found on the USDA
FDC website.  Othertimes not&#x2014;I still need to weigh a crumpet.

Food is purchased in particular units (gallons, pounds, grams).  And
in some cases the natural units are things like donuts or eggs, in
which case we may need to define our  own units (see the example of
&ldquo;xl\_egg&rdquo; below).

| Food|Quantity|Units|Price|Date|Location|FDC|
|---|---|---|---|---|---|---|
| Milk, 2% fat|1|gallon|4.99|<span class="timestamp-wrapper"><span class="timestamp">[2019-09-14 Sat]</span></span>|Monterey Market, Berkeley|336075|
| Eggs, extra large|12|xl\_egg|3.59|<span class="timestamp-wrapper"><span class="timestamp">[2019-09-14 Sat]</span></span>|Monterey Market, Berkeley|747997|
| Crumpets|6|crumpet|3.19|<span class="timestamp-wrapper"><span class="timestamp">[2019-09-14 Sat]</span></span>|Monterey Market, Berkeley|547313|
| Bananas|1|pound|3.15|<span class="timestamp-wrapper"><span class="timestamp">[2019-09-14 Sat]</span></span>|Monterey Market, Berkeley|173944|
| Carrots, Organic|2|pound|2.29|<span class="timestamp-wrapper"><span class="timestamp">[2019-09-14 Sat]</span></span>|Monterey Market, Berkeley|170393|
| Cauliflower|2.51|pound|4.24|<span class="timestamp-wrapper"><span class="timestamp">[2019-09-14 Sat]</span></span>|Monterey Market, Berkeley|169986|
| Endive, Red|1.26|pound|6.27|<span class="timestamp-wrapper"><span class="timestamp">[2019-09-14 Sat]</span></span>|Monterey Market, Berkeley|168412|
| Figs, black mission|1|pound|4.98|<span class="timestamp-wrapper"><span class="timestamp">[2019-09-14 Sat]</span></span>|Monterey Market, Berkeley|438223|
| Leeks, Organic|1|pound|1.29|<span class="timestamp-wrapper"><span class="timestamp">[2019-09-14 Sat]</span></span>|Monterey Market, Berkeley|169246|
| Lettuce, Little Gem|1|pound|5.98|<span class="timestamp-wrapper"><span class="timestamp">[2019-09-14 Sat]</span></span>|Monterey Market, Berkeley|342618|
| Mushrooms, King Oyster|1|pound|12|<span class="timestamp-wrapper"><span class="timestamp">[2019-09-14 Sat]</span></span>|Monterey Market, Berkeley|342623|
| Onion, yellow|1|pound|0.39|<span class="timestamp-wrapper"><span class="timestamp">[2019-09-14 Sat]</span></span>|Monterey Market, Berkeley|170000|
| Orange juice|0.5|gallon|8.98|<span class="timestamp-wrapper"><span class="timestamp">[2019-09-14 Sat]</span></span>|Monterey Market, Berkeley|414575|
| Parsnip|1|pound|1.98|<span class="timestamp-wrapper"><span class="timestamp">[2019-09-14 Sat]</span></span>|Monterey Market, Berkeley|170417|
| Potato, marble mix|1|pound|2.59|<span class="timestamp-wrapper"><span class="timestamp">[2019-09-14 Sat]</span></span>|Monterey Market, Berkeley|170032|
| Rhubarb|1|pound|1.84|<span class="timestamp-wrapper"><span class="timestamp">[2019-09-14 Sat]</span></span>|Monterey Market, Berkeley|167758|
| Potato, russet|10|pound|2.98|<span class="timestamp-wrapper"><span class="timestamp">[2019-09-14 Sat]</span></span>|Monterey Market, Berkeley|170030|
| Squash, Zucchini|1|pound|1.49|<span class="timestamp-wrapper"><span class="timestamp">[2019-09-14 Sat]</span></span>|Monterey Market, Berkeley|169291|



## Price information



The code below allows us to collect data on different kinds of food
with their prices from google spreadsheets.



### Using data from Google Sheets



Rather than creating & uploading `csv` files we can
also write code to find files over the internet.  Here we supply code
which allows one to input data on prices in a format similar to that
above (i.e., a spreadsheet with seven columns, labeled Food, Quantity,
Units, Price, Date, Location, and FDC).

Using google sheets in this way requires you establish some
credentials.   Follow Step 1 in [Turn on the API](https://developers.google.com/sheets/api/quickstart/python#step_1_turn_on_the_api_name).



In [7]:
# ID of sheet and name of worksheet for google sheets.   
# Easiest if these are public (or use service accounts)

SHEETs = [# Stigler's foods, modern prices
          ("https://docs.google.com/spreadsheet/ccc?key=1ObK5N_5aVXzVHE7ZXWBg0kQvPS3k1enRwsUjhytwh5A","Table 2"), 
         ]

### Compile data on food prices



In [8]:
import pandas as pd
import warnings

DFs = []
#for csv in CSVs:  # Uncomment to use a list of csv files as inputs
#  DFs.append(pd.read_csv(csv,dtype=str))

try:
    if len(SHEETs):
        for ID, RANGE_NAME in SHEETs:
           try:
               if "docs.google.com" in ID:
                   sheet = "%s&output=csv" % ID
               else:
                   sheet = "https://docs.google.com/spreadsheet/ccc?key=%s&output=csv" % ID
               DFs.append(pd.read_csv(sheet))
           except ParserError:
               warnings.warn("Can't read sheet at https://docs.google.com/spreadsheets/d/%s.\nCheck Sharing settings, so that anyone with link can view?" % ID)
except NameError: # SHEETS not defined?
    pass

df = pd.concat(DFs,ignore_index=True,sort=False)
# Some columns which ought to be numeric are actually str; convert them
df['Price'] = df['Price'].astype(float)
df['Quantity'] = df['Quantity'].astype(float)
df['FDC'] = df['FDC'].astype(int)

df

Food  Quantity   Units  Price                 Date  \
0       Wheat Flour     160.0      oz   6.41  [2019-10-15 Monday]   
1      Wheat Cereal     510.0       g   2.69  [2019-10-15 Monday]   
2         Corn Meal      24.0      oz   1.69  [2019-10-15 Monday]   
3       Rolled Oats      18.0      oz   2.99  [2019-10-15 Monday]   
4   Evaporated Milk      12.0      oz   0.94  [2019-10-15 Monday]   
5           Cabbage       1.0     lbs   0.62  [2019-10-15 Monday]   
6          Potatoes       5.0     lbs   1.47  [2019-10-15 Monday]   
7           Spinach      10.0      oz   2.38  [2019-10-15 Monday]   
8    Sweet Potatoes       1.0     lbs   1.05  [2019-10-15 Monday]   
9        Navy Beans       1.0     lbs   1.49  [2019-10-15 Monday]   
10            Sugar       4.0     lbs   2.39  [2019-10-15 Monday]   
11            Beets       1.0     lbs   2.49  [2019-10-15 Monday]   
12     Liver (Pork)       1.0     lbs   1.19  [2019-10-15 Monday]   
13     Milk (Whole)       1.0  gallon   2.65  [2019-10-15 Monday]   

            Location     FDC  
0       Deep Premium  390092  
1   Target / Kellogs  391549  
2             Target  363629  
3             Meijer  734348  
4            Walmart  648546  
5               USDA  169975  
6            Walmart  576920  
7            Walmart  168462  
8            Walmart  600987  
9             Target  535776  
10            Target  169656  
11         Good Eggs  169145  
12          ShopRite  167862  
13           Walmart  336070

### Look up nutritional information for foods



Now we have a list of foods with prices.  Do lookups on USDA database
to get nutritional information.



In [14]:
import fooddatacentral as fdc

D = {}
count = 0
for food in  df.Food.tolist():
    try:
        FDC = df.loc[df.Food==food,:].FDC[count]
        count+=1
        D[food] = fdc.nutrients(apikey[user],FDC).Quantity
    except AttributeError: 
        warnings.warn("Couldn't find FDC Code %s for food %s." % (food,FDC))        

D = pd.DataFrame(D,dtype=float)

D

Wheat Flour  Wheat Cereal  \
10:0                                              NaN           NaN   
12:0                                              NaN           NaN   
14:0                                              NaN           NaN   
14:1                                              NaN           NaN   
15:0                                              NaN           NaN   
15:1                                              NaN           NaN   
16:0                                              NaN           NaN   
16:1                                              NaN           NaN   
17:0                                              NaN           NaN   
17:1                                              NaN           NaN   
18:0                                              NaN           NaN   
18:1                                              NaN           NaN   
18:2                                              NaN           NaN   
18:3                                              NaN           NaN   
18:4                                              NaN           NaN   
20:0                                              NaN           NaN   
20:1                                              NaN           NaN   
20:2 n-6 c,c                                      NaN           NaN   
20:3                                              NaN           NaN   
20:4                                              NaN           NaN   
20:5 n-3 (EPA)                                    NaN           NaN   
22:0                                              NaN           NaN   
22:1                                              NaN           NaN   
22:5 n-3 (DPA)                                    NaN           NaN   
22:6 n-3 (DHA)                                    NaN           NaN   
24:0                                              NaN           NaN   
4:0                                               NaN           NaN   
6:0                                               NaN           NaN   
8:0                                               NaN           NaN   
Alanine                                           NaN           NaN   
...                                               ...           ...   
Sugars, total including NLEA                     6.67        51.850   
Theobromine                                       NaN           NaN   
Thiamin                                           NaN         1.000   
Threonine                                         NaN           NaN   
Tocopherol, beta                                  NaN           NaN   
Tocopherol, delta                                 NaN           NaN   
Tocopherol, gamma                                 NaN           NaN   
Tocotrienol, alpha                                NaN           NaN   
Tocotrienol, beta                                 NaN           NaN   
Tocotrienol, delta                                NaN           NaN   
Tocotrienol, gamma                                NaN           NaN   
Total lipid (fat)                               20.00         0.000   
Tryptophan                                        NaN           NaN   
Tyrosine                                          NaN           NaN   
Valine                                            NaN           NaN   
Vitamin A, IU                                    0.00      2778.000   
Vitamin A, RAE                                    NaN           NaN   
Vitamin B-12                                      NaN         5.560   
Vitamin B-12, added                               NaN           NaN   
Vitamin B-6                                       NaN         1.852   
Vitamin C, total ascorbic acid                   0.00         0.000   
Vitamin D (D2 + D3)                               NaN           NaN   
Vitamin D (D2 + D3), International Units          NaN       296.000   
Vitamin E (alpha-tocopherol)                      NaN           NaN   
Vitamin E, added                                  NaN           Na

## Units & Prices



Now, the prices we observe can be for lots of different quantities and
units.  The FDC database basically wants everything in either hundreds
of grams (hectograms) or hundreds of milliliters (deciliters).  

Sometimes this conversion is simple; if the price we observe is for
something that weighs two kilograms, that&rsquo;s just 20 hectograms.
Different systems of weights and volumes are also easy; a five pound
bag of flour is approximately 22.68 hectograms.  

Othertimes things are more complicated.  If you observe the price of a
dozen donuts, that needs to be converted to hectograms, for example.  

A function `units` in the [fdc](fooddatacentral.py) module accomplishes this conversion
for many different units, using the `python` [pint module](https://pint.readthedocs.io/en/latest/).  A file
[./Data/food\_units.txt](Data/food_units.txt) can be edited to deal with odd cases such as
donuts, using a format described in the `pint` [documentation](https://pint.readthedocs.io/en/latest/defining.html). 

Here&rsquo;s an example of the usage of `fooddatacentral.units`:



In [32]:
# Try your own quantities and units.
# If units are missing try adding to ./Data/food_units.txt

print(fdc.units(5,'lbs'))
print(fdc.units(1,'gallon'))
print(fdc.units(2,'tea_bag'))
print(fdc.units(2,'tea_bag'))

22.679618500000004 hectogram
37.85411783999999 deciliter
0.06 hectogram
0.06 hectogram


Now, use the `units` function to convert all foods to either
deciliters or hectograms, to match FDC database:



In [17]:
# Convert food quantities to FDC units
df['FDC Quantity'] = df[['Quantity','Units']].T.apply(lambda x : fdc.units(x['Quantity'],x['Units']))

# Now may want to filter df by time or place--need to get a unique set of food names.
df['FDC Price'] = df['Price']/df['FDC Quantity']

df.dropna(how='any') # Drop food with any missing data

# To use minimum price observed
Prices = df.groupby('Food')['FDC Price'].min()

Prices.head()

Food
Beets               0.5489510328403452 / hectogram
Cabbage             0.1366866025546241 / hectogram
Corn Meal          0.24838748206162872 / hectogram
Evaporated Milk    0.27631270193837987 / hectogram
Liver (Pork)       0.26235009200000425 / hectogram
Name: FDC Price, dtype: object

## Dietary Requirements



We&rsquo;ve figured out some foods we can buy, the nutritional content of
those foods, and  the price of the foods.  Now we need to say
something about nutritional requirements.   Our data for this is based
on  US government recommendations available at
[https://www.dietaryguidelines.gov/sites/default/files/2019-05/2015-2020_Dietary_Guidelines.pdf](https://www.dietaryguidelines.gov/sites/default/files/2019-05/2015-2020_Dietary_Guidelines.pdf).
Note that we&rsquo;ve tweaked the nutrient labels to match those in the FDC
data.

We&rsquo;ve broken down the requirements into three different tables.  The
first is *minimum* quantities that we need to  satisfy.  For example,
this table tells us that a 20 year-old female needs at least 46 grams
of protein per day.

| Nutrition|Source|C 1-3|F 4-8|M 4-8|F 9-13|M 9-13|F 14-18|M 14-18|F 19-30|M 19-30|F 31-50|M 31-50|F 51+|M 51+|
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
| Energy|---|1000|1200|1400|1600|1800|1800|2200|2000|2400|1800|2200|1600|2000|
| Protein|RDA|13|19|19|34|34|46|52|46|56|46|56|46|56|
| Fiber, total dietary|---|14|16.8|19.6|22.4|25.2|25.2|30.8|28|33.6|25.2|30.8|22.4|28|
| Folate, DFE|RDA|150|200|200|300|300|400|400|400|400|400|400|400|400|
| Calcium, Ca|RDA|700|1000|1000|1300|1300|1300|1300|1000|1000|1000|1000|1200|1000|
| Carbohydrate, by difference|RDA|130|130|130|130|130|130|130|130|130|130|130|130|130|
| Iron, Fe|RDA|7|10|10|8|8|15|11|18|8|18|8|8|8|
| Magnesium, Mg|RDA|80|130|130|240|240|360|410|310|400|320|420|320|420|
| Niacin|RDA|6|8|8|12|12|14|16|14|16|14|16|14|16|
| Phosphorus, P|RDA|460|500|500|1250|1250|1250|1250|700|700|700|700|700|700|
| Potassium, K|AI|3000|3800|3800|4500|4500|4700|4700|4700|4700|4700|4700|4700|4700|
| Riboflavin|RDA|0.5|0.6|0.6|0.9|0.9|1|1.3|1.1|1.3|1.1|1.3|1.1|1.3|
| Thiamin|RDA|0.5|0.6|0.6|0.9|0.9|1|1.2|1.1|1.2|1.1|1.2|1.1|1.2|
| Vitamin A, RAE|RDA|300|400|400|600|600|700|900|700|900|700|900|700|900|
| Vitamin B-12|RDA|0.9|1.2|1.2|1.8|1.8|2.4|2.4|2.4|2.4|2.4|2.4|2.4|2.4|
| Vitamin B-6|RDA|0.5|0.6|0.6|1|1|1.2|1.3|1.3|1.3|1.3|1.3|1.5|1.7|
| Vitamin C, total ascorbic acid|RDA|15|25|25|45|45|65|75|75|90|75|90|75|90|
| Vitamin E (alpha-tocopherol)|RDA|6|7|7|11|11|15|15|15|15|15|15|15|15|
| Vitamin K (phylloquinone)|AI|30|55|55|60|60|75|75|90|120|90|120|90|120|
| Zinc, Zn|RDA|3|5|5|8|8|9|11|8|11|8|11|8|11|

This next table specifies *maximum* quantities.  Our 20 year-old
female shouldn&rsquo;t have more than 2300 milligrams of sodium per day.

| Nutrition|Source|C 1-3|F 4-8|M 4-8|F 9-13|M 9-13|F 14-18|M 14-18|F 19-30|M 19-30|F 31-50|M 31-50|F 51+|M 51+|
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
| Sodium, Na|UL|1500|1900|1900|2200|2200|2300|2300|2300|2300|2300|2300|2300|2300|

Finally, we have some odd constraints given in this final table.
Mostly the items given don&rsquo;t correspond to items in the FDC data
(e.g., copper), but in some cases it may be possible to match things
up.  We can&rsquo;t use these without some additional work.

| Nutrition|Source|C 1-3|F 4-8|M 4-8|F 9-13|M 9-13|F 14-18|M 14-18|F 19-30|M 19-30|F 31-50|M 31-50|F 51+|M 51+|
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
| Carbohydrate, % kcal|AMDR|45-65|45-65|45-65|45-65|45-65|45-65|45-65|45-65|45-65|45-65|45-65|45-65|45-65|
| Added sugars, % kcal|DGA|<10%|<10%|<10%|<10%|<10%|<10%|<10%|<10%|<10%|<10%|<10%|<10%|<10%|
| Total fat, % kcal|AMDR|30-40|25-35|25-35|25-35|25-35|25-35|25-35|20-35|20-35|20-35|20-35|20-35|20-35|
| Saturated fat, % kcal|DGA|<10%|<10%|<10%|<10%|<10%|<10%|<10%|<10%|<10%|<10%|<10%|<10%|<10%|
| Linoleic acid, g|AI|7|10|10|10|12|11|16|12|17|12|17|11|14|
| Linolenic acid, g|AI|0.7|0.9|0.9|1|1.2|1.1|1.6|1.1|1.6|1.1|1.6|1.1|1.6|
| Copper, mcg|RDA|340|440|440|700|700|890|890|900|900|900|900|900|900|
| Manganese, mg|AI|1.2|1.5|1.5|1.6|1.9|1.6|2.2|1.8|2.3|1.8|2.3|1.8|2.3|
| Selenium, mcg|RDA|20|30|30|40|40|55|55|55|55|55|55|55|55|
| Choline, mg|AI|200|250|250|375|375|400|550|425|550|425|550|425|550|

-   **Notes on Source:** In each of these tables, RDA = Recommended
    Dietary Allowance, AI = Adequate Intake, UL = Tolerable Upper
    Intake Level, AMDR = Acceptable Macronutrient Distribution
    Range, DGA = 2015-2020 Dietary Guidelines recommended limit; 14
    g fiber per 1,000 kcal = basis for AI for fiber.



In [18]:
# Choose sex/age group:
group = "F 19-30"

# Define *minimums*
bmin = pd.read_csv('./diet_minimums.csv').set_index('Nutrition')[group]

# Define *maximums*
bmax = pd.read_csv('./diet_maximums.csv').set_index('Nutrition')[group]

# Inspect:
bmin

Nutrition
Energy                            2000.0
Protein                             46.0
Fiber, total dietary                28.0
Folate, DFE                        400.0
Calcium, Ca                       1000.0
Carbohydrate, by difference        130.0
Iron, Fe                            18.0
Magnesium, Mg                      310.0
Niacin                              14.0
Phosphorus, P                      700.0
Potassium, K                      4700.0
Riboflavin                           1.1
Thiamin                              1.1
Vitamin A, RAE                     700.0
Vitamin B-12                         2.4
Vitamin B-6                          1.3
Vitamin C, total ascorbic acid      75.0
Vitamin E (alpha-tocopherol)        15.0
Vitamin K (phylloquinone)           90.0
Zinc, Zn                             8.0
Name: F 19-30, dtype: float64

## Putting it together



Here we take the different pieces of the puzzle we&rsquo;ve developed and
put them together in the form of a linear program we can solve.
Recall that the mathematical problem we&rsquo;re trying to solve is
$$
    \min_x c'x
$$
such that
$$
     Ax \geq b
$$



### Objective function (c)



If we buy a bag of groceries with quantities given by $x$, the total
cost of the bag of groceries is the inner product of prices and
quantities.  Since we&rsquo;ve converted our units above, this gives us a
vector of prices where quantities are all in 100 g or ml units.



In [19]:
c = Prices.apply(lambda x:x.magnitude).dropna()

# Compile list that we have both prices and nutritional info for; drop if either missing
use = list(set(c.index.tolist()).intersection(D.columns.tolist()))
c = c[use]

c

Food
Wheat Flour        0.141316
Navy Beans         0.328489
Evaporated Milk    0.276313
Potatoes           0.064816
Milk (Whole)       0.070006
Cabbage            0.136687
Liver (Pork)       0.262350
Beets              0.548951
Sweet Potatoes     0.231485
Spinach            0.839520
Corn Meal          0.248387
Wheat Cereal       0.527451
Sugar              0.131726
Rolled Oats        0.585940
Name: FDC Price, dtype: float64

### Constraint matrix (A)



The matrix $A$ maps a bag of groceries $x$ into nutrients, but we
don&rsquo;t need to keep track of nutrients for which we don&rsquo;t have
contraints.



In [36]:
# Drop nutritional information for foods we don't know the price of,
# and replace missing nutrients with zeros.
Aall = D[c.index].fillna(0)

# Drop rows of A that we don't have constraints for.
Amin = Aall.loc[bmin.index]

Amax = Aall.loc[bmax.index]

# Minimum requirements involve multiplying constraint by -1 to make <=.
A = pd.concat([Amin,-Amax])


Wheat Flour  Navy Beans  Evaporated Milk  Potatoes  Milk (Whole)  \
Nutrition                                                                      
Sodium, Na        800.0         0.0            100.0       0.0          43.0   

            Cabbage  Liver (Pork)  Beets  Sweet Potatoes  Spinach  Corn Meal  \
Nutrition                                                                      
Sodium, Na     18.0          87.0   78.0            55.0     79.0        0.0   

            Wheat Cereal  Sugar  Rolled Oats  
Nutrition                                     
Sodium, Na          93.0    2.0          0.0

### Constraint vector ($b$)



Finally, the right hand side vector $b$ in the expression
$$
    Ax\geq b
$$



In [21]:
b = pd.concat([bmin,-bmax]) # Note sign change for max constraints

b

Nutrition
Energy                            2000.0
Protein                             46.0
Fiber, total dietary                28.0
Folate, DFE                        400.0
Calcium, Ca                       1000.0
Carbohydrate, by difference        130.0
Iron, Fe                            18.0
Magnesium, Mg                      310.0
Niacin                              14.0
Phosphorus, P                      700.0
Potassium, K                      4700.0
Riboflavin                           1.1
Thiamin                              1.1
Vitamin A, RAE                     700.0
Vitamin B-12                         2.4
Vitamin B-6                          1.3
Vitamin C, total ascorbic acid      75.0
Vitamin E (alpha-tocopherol)        15.0
Vitamin K (phylloquinone)           90.0
Zinc, Zn                             8.0
Sodium, Na                       -2300.0
Name: F 19-30, dtype: float64

## Solving the problem



First, we find a solution to the problem



In [22]:
from  scipy.optimize import linprog as lp
import numpy as np

tol = 1e-6 # Numbers in solution smaller than this (in absolute value) treated as zeros

# Now solve problem!  (Note that the linear program solver we'll use assumes
# "less-than-or-equal" constraints.  We can switch back and forth by
# multiplying $A$ and $b$ by $-1$.)

result = lp(c, -A, -b, method='interior-point')

result

     con: array([], dtype=float64)
     fun: 6.7676604210169256
 message: 'Optimization terminated successfully.'
     nit: 18
   slack: array([1.45109460e-04, 3.65972280e-06, 1.31744355e-06, 1.06942745e+03,
       1.50125167e+02, 1.61349332e+01, 1.18561809e+01, 3.19879889e+02,
       6.63544390e-01, 5.29775997e+01, 1.63954819e-04, 1.92762867e+00,
       2.69873601e-08, 4.12670467e+03, 5.93354068e+00, 9.81277596e-01,
       1.35053078e+02, 3.20891917e-08, 3.42631199e+03, 1.29361122e-05,
       1.53972067e+03])
  status: 0
 success: True
       x: array([5.71999512e-07, 4.00326009e-01, 4.79338165e-08, 2.35774724e-01,
       3.17487490e+00, 1.38662424e-06, 1.94784467e-01, 2.58623108e-08,
       7.14310379e-08, 7.27968403e+00, 7.33073850e-08, 3.31016338e-01,
       4.66706112e-01, 2.92399175e-08])

Let&rsquo;s interpret this.  Start with the cost of the solution:



In [23]:
print("Cost of diet for %s is $%4.2f per day." % (group,result.fun))

Cost of diet for F 19-30 is $6.77 per day.


Next, what is it we&rsquo;re actually eating?



In [24]:
# Put back into nice series
diet = pd.Series(result.x,index=c.index)

print("\nYou'll be eating (in 100s of grams or milliliters):")
print(diet[diet >= tol])  # Drop items with quantities less than precision of calculation.


You'll be eating (in 100s of grams or milliliters):
Food
Navy Beans      0.400326
Potatoes        0.235775
Milk (Whole)    3.174875
Cabbage         0.000001
Liver (Pork)    0.194784
Spinach         7.279684
Wheat Cereal    0.331016
Sugar           0.466706
dtype: float64


Given this diet, what are nutritional outcomes?



In [25]:
tab = pd.DataFrame({"Outcome":np.abs(A).dot(diet),"Recommendation":np.abs(b)})
print("\nWith the following nutritional outcomes of interest:")
tab


With the following nutritional outcomes of interest:


Outcome  Recommendation
Nutrition                                                  
Energy                          2000.000145          2000.0
Protein                           46.000004            46.0
Fiber, total dietary              28.000001            28.0
Folate, DFE                     1469.427446           400.0
Calcium, Ca                     1150.125167          1000.0
Carbohydrate, by difference      146.134933           130.0
Iron, Fe                          29.856181            18.0
Magnesium, Mg                    629.879889           310.0
Niacin                            14.663544            14.0
Phosphorus, P                    752.977600           700.0
Potassium, K                    4700.000164          4700.0
Riboflavin                         3.027629             1.1
Thiamin                            1.100000             1.1
Vitamin A, RAE                  4826.704666           700.0
Vitamin B-12                       8.333541             2.4
Vitamin B-6                        2.281278             1.3
Vitamin C, total ascorbic acid   210.053078            75.0
Vitamin E (alpha-tocopherol)      15.000000            15.0
Vitamin K (phylloquinone)       3516.311986            90.0
Zinc, Zn                           8.000013             8.0
Sodium, Na                       760.279333          2300.0

Finally, what are the constraints that bind?  Finding a less expensive
diet might involve finding less expensive sources for these particular nutrients.



In [26]:
print("\nConstraining nutrients are:")
excess = tab.diff(axis=1).iloc[:,1]
print(excess.loc[np.abs(excess) < tol].index.tolist())


Constraining nutrients are:
['Thiamin', 'Vitamin E (alpha-tocopherol)']
